File: java/po/world_simulator/Renderer.java
package po.world_simulator;

import javax.swing.*;
import javax.swing.border.LineBorder;

import po.world_simulator.entities.animals.Human;

import java.awt.*;
import java.awt.event.KeyEvent;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.util.LinkedList;

public class Renderer {
    protected JFrame frame;
    protected JPanel mapPanel;
    protected JPanel statusPanel;

    protected JLabel[][] gridLabels;

    protected JButton btnSave;
    protected JButton btnLoad;

    protected JLabel lblNextAction;
    protected JLabel lblCooldown;
    protected JLabel lblEntities;
    protected JTextArea logArea;

    protected LinkedList<String> logMessages = new LinkedList<>();

    public Renderer(Vector2 mapSize) {
        frame = new JFrame("World Simulator");
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.setSize(Config.WINDOW_WIDTH, Config.WINDOW_HEIGHT);
        frame.setLayout(new BorderLayout());

        // --- GLOBAL KEY LISTENER ---
        KeyboardFocusManager.getCurrentKeyboardFocusManager().addKeyEventDispatcher(new KeyEventDispatcher() {
            @Override
            public boolean dispatchKeyEvent(KeyEvent e) {
                if (e.getID() == KeyEvent.KEY_PRESSED) {
                    InputManager im = World.getInputManager();
                    if (im == null) return false;

                    int code = e.getKeyCode();
                    if (code == KeyEvent.VK_LEFT)
                        im.pushInput(Config.LEFT);
                    else if (code == KeyEvent.VK_RIGHT)
                        im.pushInput(Config.RIGHT);
                    else if (code == KeyEvent.VK_UP)
                        im.pushInput(Config.UP);
                    else if (code == KeyEvent.VK_DOWN)
                        im.pushInput(Config.DOWN);
                    else if (code == KeyEvent.VK_E)
                        im.pushInput(Config.ABILITY);
                    else if (code == KeyEvent.VK_N)
                        im.pushInput(Config.NEW_TURN);
                    else if (code == KeyEvent.VK_S)
                        im.pushInput(Config.SAVE);
                    else if (code == KeyEvent.VK_L)
                        im.pushInput(Config.LOAD);
                    else if (code == KeyEvent.VK_Q)
                        im.pushInput(Config.QUIT);
                    renderInfoWindow();
                }
                return false;
            }
        });

        // --- MAP PANEL ---
        mapPanel = new JPanel(new GridLayout(mapSize.y, mapSize.x));
        mapPanel.setBackground(Config.BG_COLOR);
        gridLabels = new JLabel[mapSize.y][mapSize.x];

        for (int y = 0; y < mapSize.y; y++) {
            for (int x = 0; x < mapSize.x; x++) {
                JLabel cell = new JLabel("", SwingConstants.CENTER);
                cell.setOpaque(true);
                cell.setBackground(Config.DEFAULT_GRID_BG_COLOR);
                cell.setBorder(new LineBorder(Config.BORDER_COLOR, Config.BORDER_THICKNESS));
                cell.setForeground(Config.TEXT_COLOR);
                cell.setFont(new Font("Monospaced", Font.BOLD, Config.BLOCK_FONT_SIZE));

                final int cx = x;
                final int cy = y;
                cell.addMouseListener(new MouseAdapter() {
                    @Override
                    public void mouseClicked(MouseEvent e) {
                        showAddEntityDialog(new Vector2(cx, cy));
                    }
                });

                gridLabels[y][x] = cell;
                mapPanel.add(cell);
            }
        }

        // --- STATUS PANEL ---
        statusPanel = new JPanel(new BorderLayout());

        int statusWidth = (Config.WINDOW_WIDTH * Config.STATUS_WIDTH) / 100;
        statusPanel.setPreferredSize(new Dimension(statusWidth, Config.WINDOW_HEIGHT));
        statusPanel.setBorder(new LineBorder(Color.BLACK, 2));

        // Górna sekcja statusu
        JPanel topStatusPanel = new JPanel();
        topStatusPanel.setLayout(new BoxLayout(topStatusPanel, BoxLayout.Y_AXIS));

        // Przyciski (B1 i B2)
        JPanel buttonsPanel = new JPanel(new FlowLayout(FlowLayout.CENTER));
        btnSave = new JButton("Save");
        btnLoad = new JButton("Load");
        btnSave.setFocusable(false); // Unikamy zabierania focusu fizycznej klawiaturze
        btnLoad.setFocusable(false);
        buttonsPanel.add(btnSave);
        buttonsPanel.add(btnLoad);

        // Informacje / Tekst
        lblNextAction = new JLabel("next action: ");
        lblCooldown = new JLabel("ability cooldown: ");
        lblEntities = new JLabel("entities: ");

        lblNextAction.setAlignmentX(Component.CENTER_ALIGNMENT);
        lblCooldown.setAlignmentX(Component.CENTER_ALIGNMENT);
        lblEntities.setAlignmentX(Component.CENTER_ALIGNMENT);

        topStatusPanel.add(buttonsPanel);
        topStatusPanel.add(lblNextAction);
        topStatusPanel.add(lblCooldown);
        topStatusPanel.add(lblEntities);

        // Dolna sekcja logów
        JPanel mainLogPanel = new JPanel(new BorderLayout());
        JLabel logTitle = new JLabel("log:", SwingConstants.CENTER);
        logArea = new JTextArea();
        logArea.setEditable(false);
        logArea.setFocusable(false);
        JScrollPane scrollPane = new JScrollPane(logArea);

        mainLogPanel.add(logTitle, BorderLayout.NORTH);
        mainLogPanel.add(scrollPane, BorderLayout.CENTER);

        statusPanel.add(topStatusPanel, BorderLayout.NORTH);
        statusPanel.add(mainLogPanel, BorderLayout.CENTER);

        frame.add(mapPanel, BorderLayout.CENTER);
        frame.add(statusPanel, BorderLayout.EAST);

        frame.setVisible(true);
    }

    // Pozwala przypisać zewnętrzną metodę, np. od InputManagera / World
    public void bind(String buttonName, Runnable action) {
        if (buttonName.equalsIgnoreCase("Save") || buttonName.equalsIgnoreCase("B1")) {
            // Czyść ewentualne stare bindy
            for (var al : btnSave.getActionListeners()) btnSave.removeActionListener(al);
            btnSave.addActionListener(e -> action.run());
        } else if (buttonName.equalsIgnoreCase("Load") || buttonName.equalsIgnoreCase("B2")) {
            for (var al : btnLoad.getActionListeners()) btnLoad.removeActionListener(al);
            btnLoad.addActionListener(e -> action.run());
        }
    }

    public void renderMap() {
        // Swing automatycznie maluje, ale może nam to odświeżyć okno, gdy zmienimy setText
        mapPanel.repaint();
    }

    public void renderInfoWindow() {
        String button_text = "next action: ";
        button_text += convertInputToText(World.getInputManager().getLastInput());
        lblNextAction.setText(button_text);

        po.world_simulator.entities.animals.Human human = Human.getInstance();

        if (human != null) {
            lblCooldown.setText("ability cooldown: " + human.getAbilityCooldown());
        } else {
            lblCooldown.setText("Human died :(");
        }

        lblEntities.setText("entities: " + World.getEntityCount());

        StringBuilder sb = new StringBuilder();
        for (String msg : logMessages) {
            sb.append(msg).append("\n");
        }
        logArea.setText(sb.toString());

        statusPanel.repaint();
    }

    public void drawCharAt(Vector2 position, char text) {
        if (position.y >= 0 && position.y < gridLabels.length &&
            position.x >= 0 && position.x < gridLabels[0].length) {
            gridLabels[position.y][position.x].setText(String.valueOf(text));
        }
    }

    public void drawMap(Vector2 mapSize) {
        for (int y = 0; y < mapSize.y; y++) {
            for (int x = 0; x < mapSize.x; x++) {
                gridLabels[y][x].setText("");
            }
        }
    }

    public void addToLog(String text) {
        logMessages.add(text);
        // Zabezpieczenie by log się nie rozrastał w nieskończoność
        if (logMessages.size() > 100) {
            logMessages.removeFirst();
        }
    }

    public void drawInfoWindow() {
        // Logikę realizuje już złączone na stałe update w renderInfoWindow() dla Swinga
    }

    public void clearLog() {
        logMessages.clear();
        logArea.setText("");
    }

    public void dispose() {
        frame.dispose();
    }


    public String convertInputToText(int input){
    switch (input){
    case Config.LEFT:
        return "LEFT";
    case Config.RIGHT:
        return "RIGHT";
    case Config.UP:
        return "UP";
    case Config.DOWN:
        return "DOWN";
    case Config.ABILITY:
        return "ABILITY";
    default:
        return "";
    }
    }

    protected void showAddEntityDialog(Vector2 position) {
        po.world_simulator.Map map = World.getMap();
        if (map != null && map.isTileOccupied(position)) {
            return;
        }

        String[] entities = {
            "Wolf", "Sheep", "Fox", "Turtle", "Antelope",
            "Grass", "Milkweed", "Guarana", "Wolfberries", "SosnowskiHogweed"
        };

        String message = "Pozycja pola: (" + position.x + ", " + position.y + ")\nWybierz obiekt do dodania:";
        String selected = (String) JOptionPane.showInputDialog(
            frame,
            message,
            "Dodaj Entity",
            JOptionPane.PLAIN_MESSAGE,
            null,
            entities,
            entities[0]
        );

        if (selected != null && !selected.isEmpty()) {
            po.world_simulator.entities.Entity newEntity = null;
            switch (selected) {
                case "Wolf": newEntity = new po.world_simulator.entities.animals.Wolf(position); break;
                case "Sheep": newEntity = new po.world_simulator.entities.animals.Sheep(position); break;
                case "Fox": newEntity = new po.world_simulator.entities.animals.Fox(position); break;
                case "Turtle": newEntity = new po.world_simulator.entities.animals.Turtle(position); break;
                case "Antelope": newEntity = new po.world_simulator.entities.animals.Antelope(position); break;
                case "Grass": newEntity = new po.world_simulator.entities.plants.Grass(position); break;
                case "Milkweed": newEntity = new po.world_simulator.entities.plants.Milkweed(position); break;
                case "Guarana": newEntity = new po.world_simulator.entities.plants.Guarana(position); break;
                case "Wolfberries": newEntity = new po.world_simulator.entities.plants.Wolfberries(position); break;
                case "SosnowskiHogweed": newEntity = new po.world_simulator.entities.plants.SosnowskiHogweed(position); break;
            }

            if (newEntity != null) {
                World.getInstance().addNewEntity(newEntity);
                addToLog("Added " + selected + " at (" + position.x + ", " + position.y + ")");
                World.getInstance().forceDrawAll();
            }
        }
    }
}

File: python/po/world_simulator/Renderer.py
import tkinter as tk
from tkinter import simpledialog
import collections

from po.world_simulator.Config import Config
from po.world_simulator.Vector2 import Vector2
import po.world_simulator.World as world_module
from po.world_simulator.entities.animals.Human import Human

# Helper function to convert custom Color class to tkinter hex colors
def get_hex_color(color):
    return f"#{color.r:02x}{color.g:02x}{color.b:02x}"

class Renderer:
    def __init__(self, mapSize):
        self.frame = tk.Tk()
        self.frame.title("World Simulator")
        self.frame.geometry(f"{Config.WINDOW_WIDTH}x{Config.WINDOW_HEIGHT}")
        self.frame.protocol("WM_DELETE_WINDOW", self.on_close)

        self.logMessages = collections.deque()
        self.mapSize = mapSize

        # --- GLOBAL KEY LISTENER ---
        self.frame.bind("<KeyPress>", self.dispatchKeyEvent)

        # main container
        self.main_container = tk.Frame(self.frame)
        self.main_container.pack(fill=tk.BOTH, expand=True)

        # --- MAP PANEL ---
        self.mapPanel = tk.Frame(self.main_container, bg=get_hex_color(Config.BG_COLOR))
        self.mapPanel.pack(side=tk.LEFT, fill=tk.BOTH, expand=True)

        self.gridLabels = [[None for _ in range(mapSize.x)] for _ in range(mapSize.y)]

        # Tkinter grid configuration to make cells expand equally
        for y in range(mapSize.y):
            self.mapPanel.rowconfigure(y, weight=1)
        for x in range(mapSize.x):
            self.mapPanel.columnconfigure(x, weight=1)

        for y in range(mapSize.y):
            for x in range(mapSize.x):
                # Using frame for different border thickness
                border_frame = tk.Frame(self.mapPanel, bg=get_hex_color(Config.BORDER_COLOR), bd=Config.BORDER_THICKNESS)
                border_frame.grid(row=y, column=x, sticky="nsew")

                cell = tk.Label(border_frame, text="", bg=get_hex_color(Config.DEFAULT_GRID_BG_COLOR),
                                fg=get_hex_color(Config.TEXT_COLOR), font=("Courier", Config.BLOCK_FONT_SIZE, "bold"))
                cell.pack(fill=tk.BOTH, expand=True)

                cell.bind("<Button-1>", lambda e, cx=x, cy=y: self.showAddEntityDialog(Vector2(cx, cy)))

                self.gridLabels[y][x] = cell

        # --- STATUS PANEL ---
        statusWidth = int((Config.WINDOW_WIDTH * Config.STATUS_WIDTH) / 100)
        self.statusPanel = tk.Frame(self.main_container, width=statusWidth, highlightbackground="black", highlightthickness=2)
        self.statusPanel.pack(side=tk.RIGHT, fill=tk.Y)
        self.statusPanel.pack_propagate(False) # Prevent shrinking

        # upper status panel
        topStatusPanel = tk.Frame(self.statusPanel)
        topStatusPanel.pack(side=tk.TOP, fill=tk.X, pady=10)

        # buttons
        buttonsPanel = tk.Frame(topStatusPanel)
        buttonsPanel.pack(pady=5)

        # Buttons, disabling strict focus so key events still propagate to root
        self.btnSave = tk.Button(buttonsPanel, text="Save", takefocus=False)
        self.btnSave.pack(side=tk.LEFT, padx=5)
        self.btnLoad = tk.Button(buttonsPanel, text="Load", takefocus=False)
        self.btnLoad.pack(side=tk.LEFT, padx=5)

        # info
        self.lblNextAction = tk.Label(topStatusPanel, text="next action: ")
        self.lblNextAction.pack()
        self.lblCooldown = tk.Label(topStatusPanel, text="ability cooldown: ")
        self.lblCooldown.pack()
        self.lblEntities = tk.Label(topStatusPanel, text="entities: ")
        self.lblEntities.pack()

        # log area
        mainLogPanel = tk.Frame(self.statusPanel)
        mainLogPanel.pack(side=tk.BOTTOM, fill=tk.BOTH, expand=True)

        logTitle = tk.Label(mainLogPanel, text="log:")
        logTitle.pack(side=tk.TOP)

        self.logArea = tk.Text(mainLogPanel, state=tk.DISABLED, takefocus=False)
        scrollPane = tk.Scrollbar(mainLogPanel, command=self.logArea.yview)
        self.logArea.configure(yscrollcommand=scrollPane.set)

        self.logArea.pack(side=tk.LEFT, fill=tk.BOTH, expand=True)
        scrollPane.pack(side=tk.RIGHT, fill=tk.Y)

        # Keep the window responsive for setup but wait for manual update calls during simulation
        self.frame.update()


    def on_close(self):
        # Quit cleanly when clicking X
        im = world_module.World.getInputManager()
        if im is not None:
            im.pushInput(Config.QUIT) # Config.QUIT


    def dispatchKeyEvent(self, e):
        im = world_module.World.getInputManager()
        if im is None:
            return "break"

        keysym = e.keysym.upper()
        # Map tkinter keysyms to our input setup
        
        im.pushInput(keysym.lower())

        return "break"


    def bind(self, buttonName, action):
        if buttonName.lower() == "save" or buttonName.lower() == "b1":
            self.btnSave.config(command=action)
        elif buttonName.lower() == "load" or buttonName.lower() == "b2":
            self.btnLoad.config(command=action)


    def renderMap(self):
        self.frame.update()


    def renderInfoWindow(self):
        button_text = "next action: "
        button_text += self.convertInputToText(world_module.World.getInputManager().getLastInput())
        self.lblNextAction.config(text=button_text)

        human = Human.getInstance()

        if human is not None:
            self.lblCooldown.config(text=f"ability cooldown: {human.getAbilityCooldown()}")
        else:
            self.lblCooldown.config(text="Human died :(")

        self.lblEntities.config(text=f"entities: {world_module.World.getEntityCount()}")

        sb = ""
        for msg in self.logMessages:
            sb += msg + "\n"

        self.logArea.config(state=tk.NORMAL)
        self.logArea.delete(1.0, tk.END)
        self.logArea.insert(tk.END, sb)
        self.logArea.config(state=tk.DISABLED)
        # self.logArea.yview(tk.END) # autoscroll optionally

        self.frame.update()


    def drawCharAt(self, position, text):
        if 0 <= position.y < len(self.gridLabels) and 0 <= position.x < len(self.gridLabels[0]):
            self.gridLabels[position.y][position.x].config(text=str(text))


    def drawMap(self, mapSize):
        for y in range(mapSize.y):
            for x in range(mapSize.x):
                self.gridLabels[y][x].config(text=" ")


    def addToLog(self, text):
        if len(self.logMessages) <= Config.MAX_LOG_MESSAGES:
            self.logMessages.append(text)


    def drawInfoWindow(self):
        pass


    def clearLog(self):
        self.logMessages.clear()
        self.logArea.config(state=tk.NORMAL)
        self.logArea.delete(1.0, tk.END)
        self.logArea.config(state=tk.DISABLED)


    def dispose(self):
        self.frame.destroy()


    def convertInputToText(self, input_val):
        if input_val == Config.LEFT:
            return "LEFT"
        elif input_val == Config.RIGHT:
            return "RIGHT"
        elif input_val == Config.UP:
            return "UP"
        elif input_val == Config.DOWN:
            return "DOWN"
        elif input_val == Config.ABILITY:
            return "ABILITY"
        return ""


    def showAddEntityDialog(self, position):
        game_map = world_module.World.getMap()
        if game_map is not None and game_map.isTileOccupied(position):
            return

        entities_list = [
            "Wolf", "Sheep", "Fox", "Turtle", "Antelope",
            "Grass", "Milkweed", "Guarana", "Wolfberries", "SosnowskiHogweed", "CyberSheep"
        ]

        # Using custom dialog with OptionMenu to select entity
        dialog = tk.Toplevel(self.frame)
        dialog.title("Add Entity")
        dialog.transient(self.frame)
        dialog.grab_set()

        tk.Label(dialog, text=f"Position: ({position.x}, {position.y})\nChoose entity to add:").pack(pady=10, padx=20)

        selected_entity = tk.StringVar(dialog)
        selected_entity.set(entities_list[0])

        tk.OptionMenu(dialog, selected_entity, *entities_list).pack(pady=5, padx=20)

        result = [None]

        def on_ok():
            result[0] = selected_entity.get()
            dialog.destroy()

        def on_cancel():
            dialog.destroy()

        btn_frame = tk.Frame(dialog)
        btn_frame.pack(pady=10)
        tk.Button(btn_frame, text="OK", command=on_ok, width=10).pack(side=tk.LEFT, padx=5)
        tk.Button(btn_frame, text="Cancel", command=on_cancel, width=10).pack(side=tk.LEFT, padx=5)

        # Center the dialog relative to the main window
        dialog.update_idletasks()
        x = self.frame.winfo_x() + (self.frame.winfo_width() // 2) - (dialog.winfo_width() // 2)
        y = self.frame.winfo_y() + (self.frame.winfo_height() // 2) - (dialog.winfo_height() // 2)
        dialog.geometry(f"+{x}+{y}")

        self.frame.wait_window(dialog)
        selected = result[0]

        # Validate selected string
        if selected and selected in entities_list:
            newEntity = None
            try:
                # Same strategy as World string loader
                if selected in ["Wolf", "Sheep", "Fox", "Turtle", "Antelope", "Human", "CyberSheep"]:
                    import importlib
                    module = importlib.import_module(f"po.world_simulator.entities.animals.{selected}")
                    clazz = getattr(module, selected)
                else:
                    import importlib
                    module = importlib.import_module(f"po.world_simulator.entities.plants.{selected}")
                    clazz = getattr(module, selected)

                newEntity = clazz(position)
            except Exception as e:
                print("Failed to create entity directly from dialog:", e)

            if newEntity is not None:
                world_module.World.getInstance().addNewEntity(newEntity)
                self.addToLog(f"Added {selected} at ({position.x}, {position.y})")
                world_module.World.getInstance().forceDrawAll()
